File: Chapter 02/C#/2.3.4 Performance/CacheService.cs
using System;
using System.Collections.Generic;

namespace Chapter02.Performance;

public class CacheService
{
    private readonly Dictionary<string, object> _store = new();

    public object? Get(string key)
    {
        Console.WriteLine($"  [CACHE] Checking for key: {key}");
        if (_store.TryGetValue(key, out var value))
        {
            Console.WriteLine("  [CACHE] HIT! Returning data immediately. (takes 5ms)");
            return value;
        }
        Console.WriteLine("  [CACHE] MISS!");
        return null;
    }

    public void Set(string key, object value, int ttlSeconds)
    {
        Console.WriteLine($"  [CACHE] Saving data for key: {key} (Expires in {ttlSeconds}s)");
        _store[key] = value;
    }
}

File: Chapter 02/JS/2.3.4 Performance/cacheService.js
const store = new Map();

const cache = {
    get: (key) => {
        console.log(`  [CACHE] Checking for key: ${key}`);
        if (store.has(key)) {
            console.log("  [CACHE] HIT! Returning data immediately. (takes 5ms)");
            return store.get(key);
        }
        console.log("  [CACHE] MISS!");
        return null;
    },
    set: (key, value, ttlSeconds) => {
        console.log(`  [CACHE] Saving data for key: ${key} (Expires in ${ttlSeconds}s)`);
        store.set(key, value);
    }
};

module.exports = cache;